### PR TITLE
Fix some minor scheduler issues with groups

### PR DESCRIPTION
**Pull request by kgaillot**
The scheduler group code is full of corner cases that are not correctly handled. This is the start of a project to clean them up.

It starts by adding four new regression tests, one of which indicates a problem that is not corrected in this PR but hopefully will be separately.

### CODE REVIEW

**Comment by kgaillot** [comment]
@nrwahl2 , can you review this when you get a chance? Thanks ...

File: lib/pacemaker/pcmk_sched_colocation.c [hunk 1]
```diff
@@ -1265,6 +1265,43 @@ init_group_colocated_nodes(const pe_resource_t *rsc, const char *log_id,
     return work;
 }
 
+/*!
+ * \internal
+ * \brief Initialize colocated node table for a non-group resource
+ *
+ * \param[in] rsc     Group resource being colocated with another resource
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_colocation.c:1303]: not group

File: lib/pacemaker/pcmk_sched_colocation.c [hunk 2]
```diff
@@ -1265,6 +1265,43 @@ init_group_colocated_nodes(const pe_resource_t *rsc, const char *log_id,
...
+ * \param[in] log_id  Resource ID to use in log messages
+ * \param[in] nodes   Nodes to update
+ * \param[in] attr    Colocation attribute (NULL to use default)
+ * \param[in] factor  Incorporate scores multiplied by this factor
+ * \param[in] flags   Bitmask of enum pcmk__coloc_select values
+ *
+ * \return Table of node scores initialized for colocation
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_colocation.c]: ditto on maybe adding a free-ing note

File: lib/pacemaker/pcmk_sched_colocation.c [hunk 3]
```diff
@@ -1220,6 +1220,51 @@ add_node_scores_matching_attr(GHashTable *nodes, const pe_resource_t *rsc,
...
+init_group_colocated_nodes(const pe_resource_t *rsc, const char *log_id,
+                           GHashTable **nodes, const char *attr, float factor,
+                           uint32_t flags)
+{
+    GHashTable *work = NULL;
+    pe_resource_t *member = NULL;
+
+    if (*nodes == NULL) {
+        // Only cmp_resources() passes a NULL nodes table
+        member = pe__last_group_member(rsc);
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_colocation.c:1277]: This is probably correct and it matches the existing logic, but the reasoning for choosing the last resource instead of the first in this case isn't immediately obvious to me. I haven't had time to do a deeper dive yet.

Your commit message in an earlier commit says that `pcmk__group_assign()` adds the group's colocations to the last member. But `cmp_resources()` appears to run before the assign function, so assign hasn't been called yet.

File: lib/pacemaker/pcmk_sched_colocation.c [hunk 4]
```diff
@@ -130,6 +130,34 @@ cmp_primary_priority(gconstpointer a, gconstpointer b)
     return strcmp(rsc_constraint1->primary->id, rsc_constraint2->primary->id);
 }
 
+/*!
+ * \internal
+ * \brief Add a "this with" colocation constraint to a resource
+ *
+ * \param[in,out] rsc         Resource to add colocation to
+ * \param[in]     colocation  Colocation constraint to add
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_colocation.c:138]: If we say "to \p rsc" in the "with this" function, we might as well here

File: lib/pacemaker/pcmk_sched_group.c [hunk 5]
```diff
@@ -61,20 +64,23 @@ expand_group_colocations(pe_resource_t *rsc)
...
-                    member->rsc_cons = g_list_prepend(member->rsc_cons, constraint);
+                    pcmk__add_this_with(member, constraint);
                 }
             }
         } else if (!pcmk_is_set(member->flags, pe_rsc_managed)) {
             any_unmanaged = true;
         }
     }
 
+    g_list_free(rsc->rsc_cons);
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_group.c:75]: This only frees the list and doesn't touch the elements, right? AFAICT only `g_list_free_full()` frees the elements, but the doc says "The freed elements are returned to the slice allocator." Not sure off hand what that means.

https://docs.gtk.org/glib/type_func.List.free.html

**Review by nrwahl2 — COMMENTED** [review]
Looks great

File: cts/cts-scheduler.in [hunk 6]
```diff
@@ -235,6 +235,7 @@ TESTS = [
         [ "anti-colocation-promoted", "Organize order of actions for promoted resources in anti-colocations" ],
         [ "anti-colocation-unpromoted", "Organize order of actions for unpromoted resources in anti-colocations" ],
         [ "group-anticolocation", "Group with failed last member anti-colocated with another group" ],
+        [ "group-colocation-failure", "Group with failed last member colocated with another group" ],
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply cts/cts-scheduler.in:238]: Good point, it is specific to single-member groups

File: cts/scheduler/scores/colocation-priority-group.scores [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+pcmk__group_assign: group1 allocation score on node1: 0
+pcmk__group_assign: group1 allocation score on node2: 0
+pcmk__group_assign: member1a allocation score on node1: 0
+pcmk__group_assign: member1a allocation score on node2: 0
+pcmk__group_assign: member1b allocation score on node1: 0
+pcmk__group_assign: member1b allocation score on node2: 0
+pcmk__primitive_assign: Fencing allocation score on node1: 0
+pcmk__primitive_assign: Fencing allocation score on node2: 0
+pcmk__primitive_assign: member1a allocation score on node1: 0
+pcmk__primitive_assign: member1a allocation score on node2: -4000
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply cts/scheduler/scores/colocation-priority-group.scores:11]: Yeah, in pcmk__apply_coloc_to_weights(), if the primary isn't running, the negative of the score is applied to all the dependent's allowed nodes. Since it's the same for all nodes, presumably it would matter only if INFINITY. (Anti-colocations have already been filtered out by that point.)

File: cts/cts-scheduler.in [hunk 8]
```diff
@@ -242,6 +242,7 @@ TESTS = [
         [ "coloc-dependee-should-move", "Group colocation outweighs stickiness" ],
         [ "colocation-influence", "Respect colocation influence" ],
         [ "colocation-priority-group", "Apply group colocations in order of primary priority" ],
+        [ "colocation-vs-stickiness", "Group stickiness outweighs anti-colocation score" ],
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply cts/cts-scheduler.in:245]: The test description isn't intended to mean that stickiness always outweighs anti-colocations, but that stickiness should win with this particular input because its score is higher.

Clarifying the docs is a good idea

File: lib/pacemaker/pcmk_sched_colocation.c [hunk 9]
```diff
@@ -1220,6 +1220,51 @@ add_node_scores_matching_attr(GHashTable *nodes, const pe_resource_t *rsc,
...
+ * \brief Initialize colocated node table for a group resource
+ *
+ * \param[in] rsc     Group resource being colocated with another resource
+ * \param[in] log_id  Resource ID to use in log messages
+ * \param[in] nodes   Nodes to update
+ * \param[in] attr    Colocation attribute (NULL to use default)
+ * \param[in] factor  Incorporate scores multiplied by this factor
+ * \param[in] flags   Bitmask of enum pcmk__coloc_select values
+ *
+ * \return Table of node scores initialized for colocation
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_colocation.c]: Might as well

File: lib/pacemaker/pcmk_sched_colocation.c (hunk 1 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_colocation.c:1303]: whoops

File: lib/pacemaker/pcmk_sched_colocation.c (hunk 3 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_colocation.c:1277]: I hadn't gotten as far as dealing with that yet. I suspect it has issues. The PR was going to be scary long if I kept going :)

File: lib/pacemaker/pcmk_sched_group.c (hunk 5 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_group.c:75]: Correct, it's a little obscure here: pe_working_set_t has a complete list of all colocation constraints in the cluster, then each resource has the constraints affecting it added to its rsc_cons or rsc_cons_lhs. Therefore, we free only the lists, not the constraints themselves, when freeing rsc_cons and rsc_cons_lhs; the constraints are freed when the pe_working_set_t is freed.

**Comment by kgaillot** [comment]
Updated per review:

* Description for group-colocation-failure test updated
* Added documentation commit on top
* Updated doxygen blocks for init\_\*\_colocated_nodes() and pcmk\_\_add\_this\_with()

File: cts/scheduler/scores/colocation-priority-group.scores (hunk 7 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply cts/scheduler/scores/colocation-priority-group.scores:11]: Ah okay, so it sounds like the 0 is from 4000 + (-4000) then too. (crm_simulate -VVVVVV probably would have told me that)

File: doc/sphinx/Pacemaker_Explained/constraints.rst [hunk 10]
```diff
@@ -585,15 +585,13 @@ Advisory Placement
...
-becomes mandatory, advisory colocation constraints can combine with
-other elements of the configuration to behave as if they were
-mandatory.
+placement is the "I'd prefer if" alternative.
+
+For colocation constraints with scores greater than **-INFINITY** and less than
+**INFINITY**, the cluster will try to accommodate your wishes, but may ignore
+them if other factors outweigh the colocation score. Those factors might
+include other constraints, resource stickiness, failure thresholds, whether
+other resources would be prevented from being active, etc.
```
Comment (nrwahl2) [inline doc/sphinx/Pacemaker_Explained/constraints.rst:594]: Much clearer :)

**Review by nrwahl2 — APPROVED** [review]
